### PR TITLE
Worker activity server side pagination

### PR DESCRIPTION
**Pull request by emord**
First commit turns n ES queries into 1 (n is user)
Second filters query data by user ids

cc @gcapalbo

### CODE REVIEW

File: corehq/apps/reports/standard/monitoring.py [hunk 1]
```diff
@@ -1494,12 +1495,12 @@ def _report_data(self):
...
         return WorkerActivityReportData(
-            avg_submissions_by_user=get_submission_counts_by_user(self.domain, avg_datespan),
-            submissions_by_user=get_submission_counts_by_user(self.domain, self.datespan),
-            active_cases_by_owner=get_active_case_counts_by_owner(self.domain, self.datespan, self.case_types),
-            total_cases_by_owner=get_total_case_counts_by_owner(self.domain, self.datespan, self.case_types),
-            cases_closed_by_user=get_case_counts_closed_by_user(self.domain, self.datespan, self.case_types),
-            cases_opened_by_user=get_case_counts_opened_by_user(self.domain, self.datespan, self.case_types),
+            avg_submissions_by_user=get_submission_counts_by_user(self.domain, avg_datespan, self.user_ids),
+            submissions_by_user=get_submission_counts_by_user(self.domain, self.datespan, self.user_ids),
+            active_cases_by_owner=get_active_case_counts_by_owner(self.domain, self.datespan, self.case_types, self.user_ids),
```
Comment (dimagimon) [inline corehq/apps/reports/standard/monitoring.py]: E501 line too long (126 > 115 characters)

File: corehq/apps/reports/standard/monitoring.py [hunk 2]
```diff
@@ -1494,12 +1495,12 @@ def _report_data(self):
...
+            total_cases_by_owner=get_total_case_counts_by_owner(self.domain, self.datespan, self.case_types, self.user_ids),
+            cases_closed_by_user=get_case_counts_closed_by_user(self.domain, self.datespan, self.case_types, self.user_ids),
```
Comment (dimagimon) [inline corehq/apps/reports/standard/monitoring.py]: E501 line too long (124 > 115 characters)

File: corehq/apps/reports/standard/monitoring.py [hunk 3]
```diff
@@ -1494,12 +1495,12 @@ def _report_data(self):
...
-            avg_submissions_by_user=get_submission_counts_by_user(self.domain, avg_datespan),
-            submissions_by_user=get_submission_counts_by_user(self.domain, self.datespan),
-            active_cases_by_owner=get_active_case_counts_by_owner(self.domain, self.datespan, self.case_types),
-            total_cases_by_owner=get_total_case_counts_by_owner(self.domain, self.datespan, self.case_types),
-            cases_closed_by_user=get_case_counts_closed_by_user(self.domain, self.datespan, self.case_types),
-            cases_opened_by_user=get_case_counts_opened_by_user(self.domain, self.datespan, self.case_types),
+            avg_submissions_by_user=get_submission_counts_by_user(self.domain, avg_datespan, self.user_ids),
+            submissions_by_user=get_submission_counts_by_user(self.domain, self.datespan, self.user_ids),
+            active_cases_by_owner=get_active_case_counts_by_owner(self.domain, self.datespan, self.case_types, self.user_ids),
+            total_cases_by_owner=get_total_case_counts_by_owner(self.domain, self.datespan, self.case_types, self.user_ids),
```
Comment (dimagimon) [inline corehq/apps/reports/standard/monitoring.py]: E501 line too long (124 > 115 characters)

File: corehq/apps/reports/standard/monitoring.py [hunk 4]
```diff
@@ -1494,12 +1495,12 @@ def _report_data(self):
...
+            cases_closed_by_user=get_case_counts_closed_by_user(self.domain, self.datespan, self.case_types, self.user_ids),
+            cases_opened_by_user=get_case_counts_opened_by_user(self.domain, self.datespan, self.case_types, self.user_ids),
```
Comment (dimagimon) [inline corehq/apps/reports/standard/monitoring.py]: E501 line too long (124 > 115 characters)

File: corehq/apps/reports/analytics/esaccessors.py [hunk 5]
```diff
@@ -20,30 +20,47 @@
...
-        .sort("form.meta.timeEnd", desc=True) \
-        .size(1)
-    results = form_query.run().hits
-
+def get_last_submission_time_for_user(domain, user_ids, datespan):
     def convert_to_date(date):
         return string_to_datetime(date).date() if date else None
 
-    return convert_to_date(results[0]['form']['meta']['timeEnd'] if results else None)
+    query = (
```
Comment (emord) [inline corehq/apps/reports/analytics/esaccessors.py]: This is mostly a simplified version of https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/reports/analytics/esaccessors.py#L131-L176 that only returns `{user_id: timeEnd}`

**Comment by benrudolph** [comment]
yay for aggregations and reducing the load on our ES cluster! 💪

File: corehq/apps/reports/analytics/esaccessors.py [hunk 6]
```diff
@@ -20,19 +20,36 @@
...
+        .user_id(user_ids)
+        .completed(gte=datespan.startdate.date(), lte=datespan.enddate.date())
+        .aggregation(
+            TermsAggregation('user_id', 'form.meta.userID').aggregation(
+                TopHitsAggregation(
+                    'top_hits_last_form_submissions',
+                    'form.meta.timeEnd',
+                    is_ascending=False,
+                    include='form.meta.timeEnd',
+                )
```
Comment (snopoke) [inline corehq/apps/reports/analytics/esaccessors.py]: nice

File: corehq/apps/reports/analytics/esaccessors.py [hunk 7]
```diff
@@ -20,19 +20,36 @@
...
-def get_last_submission_time_for_user(domain, user_id, datespan):
-    form_query = FormES() \
-        .domain(domain) \
-        .user_id([user_id]) \
-        .completed(gte=datespan.startdate.date(), lte=datespan.enddate.date()) \
-        .sort("form.meta.timeEnd", desc=True) \
-        .size(1)
-    results = form_query.run().hits
-
+def get_last_submission_time_for_user(domain, user_ids, datespan):
```
Comment (snopoke) [inline corehq/apps/reports/analytics/esaccessors.py]: name should be `...for_users`

File: corehq/apps/reports/analytics/esaccessors.py [hunk 8]
```diff
@@ -107,6 +110,9 @@ def _get_case_counts_by_user(domain, datespan, case_types=None, is_opened=True):
     if case_types:
         case_query = case_query.filter({"terms": {"type.exact": case_types}})
 
+    if owner_ids:
+        case_query = case_query.filter({"terms": {user_field: owner_ids}})
```
Comment (snopoke) [inline corehq/apps/reports/analytics/esaccessors.py]: could use `filters.term(user_field, owner_ids)`

File: corehq/apps/reports/analytics/esaccessors.py [hunk 9]
```diff
@@ -107,6 +110,9 @@ def _get_case_counts_by_user(domain, datespan, case_types=None, is_opened=True):
     if case_types:
         case_query = case_query.filter({"terms": {"type.exact": case_types}})
```
Comment (snopoke) [inline corehq/apps/reports/analytics/esaccessors.py]: Not you're code I know but could be `case_query.case_type(case_types)`

**Comment by snopoke** [comment]
merging but please follow up on comments

**Comment by czue** [comment]
:+1: this is great thanks

**Pull request merged**